src/main.mligo

```diff
@@ -0,0 +1,154 @@
+        let (pub_key, (sig, hash_)) = permit in
+        let packed = Bytes.pack (((Tezos.get_chain_id()), Tezos.get_self_address()), (ext.counter, hash_)) in
+        if Crypto.check pub_key sig packed
+        then
+            let sender_ = Tezos.address (Tezos.implicit_account (Crypto.hash_key pub_key)) in
+            let permit_key = sender_, hash_ in
+            match Big_map.find_opt permit_key ext.permits with
+            | None -> Extension.add_permit ext permit_key
+            | Some submission_timestamp ->
+                let () = Extension._check_not_expired s.extension submission_timestamp permit_key in
+                Extension.update_permit ext permit_key
+        else ([%Michelson ({| { FAILWITH } |} : string * bytes -> Extension.t)]) (Errors.missigned, packed)
+    in
+    let extension = List.fold_left process_permit s.extension lst in
+    { s with extension = extension }
+
+(* TZIP-17 *)
+let set_expiry (p : expiry_params) (s : storage) =
+    let (user_address, (seconds, permit_hash_opt)) = p in
+    if seconds > s.extension.max_expiry
+    then (failwith Errors.max_seconds_exceeded : storage)
+    else if Tezos.get_sender() <> user_address
+    then (failwith Errors.forbidden_expiry_update : storage)
+    else
+        let new_storage =
+            match permit_hash_opt with
+            | None ->
+                {
+                    s with extension.user_expiries = Big_map.add
+                        user_address
+                        (Some seconds)
+                        s.extension.user_expiries
+                }
+            | Some permit_hash ->
+                {
+                    s with extension.permit_expiries = Big_map.add
+                        (user_address, permit_hash)
+                        (Some seconds)
+                        s.extension.permit_expiries
+                }
+        in new_storage
+
+(* TZIP-17 implementation of TZIP-12 Transfer *)
+let transfer_permitted (t:FA2.transfer) (s: storage) =
+    let make_transfer (acc, t : (FA2.Ledger.t * Extension.t) * FA2.transfer_from) =
+        let (l, e) = acc in
+        let (is_transfer_authorized, e) = Extension.transfer_presigned e t in
+        List.fold
+          (fun (acc, dst : (FA2.Ledger.t * Extension.t) * FA2.atomic_trans) ->
+            let () = FA2.Storage.assert_token_exist s dst.token_id in
+            let () = if not is_transfer_authorized then
+                FA2.Operators.assert_authorisation s.operators t.from_ dst.token_id
+            in
+            let (ledger, ext) = acc in
+            let ledger = FA2.Ledger.decrease_token_amount_for_user ledger t.from_ dst.token_id dst.amount in
+            let ledger = FA2.Ledger.increase_token_amount_for_user ledger dst.to_ dst.token_id dst.amount in
+            (ledger, ext)
+          ) t.tx (l, e)
+        in
+    let (new_ledger, new_ext) = List.fold make_transfer t (s.ledger, s.extension)
+    in { s with ledger = new_ledger; extension = new_ext }
```

Comment:
```suggestion
let transfer_permitted (t:FA2.transfer) (s: storage) =
    let make_transfer (acc, t : (FA2.Ledger.t * Extension.t) * FA2.transfer_from) =
        let (ledger, ext) = acc in
        let (is_transfer_authorized, ext) = Extension.transfer_presigned ext t in
        let {from_; tx} = t in
        let ledger = List.fold
          (fun (ledger, dst : FA2.Ledger.t * FA2.atomic_trans) ->
            let {token_id; amount; to_} = dst in
            let () = FA2.Storage.assert_token_exist s token_id in
            let () = if not is_transfer_authorized then
                FA2.Operators.assert_authorisation s.operators from_ token_id
            in
            let ledger = FA2.Ledger.decrease_token_amount_for_user ledger from_ token_id amount in
            let ledger = FA2.Ledger.increase_token_amount_for_user ledger to_ token_id amount in
            ledger
          ) tx ledger in
          (ledger, ext)
        in
    let (new_ledger, new_ext) = List.fold make_transfer t (s.ledger, s.extension)
    in { s with ledger = new_ledger; extension = new_ext }
```
Some minor changes that leads to smaller michelson

Reply:
thanks for the best practice tip, will try to remember